### ISSUE TITLE
[dv] time based dumps?

### ISSUE THREAD

**Issue by tjaychen** [lowRISC/opentitan#15361]
should we consider adding dvsim options to control when dumps should turn on/off?
Some of the longer sims (especially those involving bootstrap), don't always have something interesting to debug until a certain point in time. 

I took a look around, and i don't really if the right answer is to add another dvsim argument and pass it all the way through (as environment variables to sim.tcl?), or if there's a smarter way.

**Comment by sriyerg** [comment]
For now, you can uncomment `sim.tcl` to set `dump_tb_top` to 0. 
https://github.com/lowRISC/opentitan/blob/488fd246967f7916220fc04099f562924804b9c7/hw/dv/tools/sim.tcl#L23

Then, modify line 44 to do:
```
  run 10us 
  wavedumpScope $waves $simulator $tb_top 0
  run
```

**Comment by cdgori** [comment]
Ultimately it would be nice to have this (it can really speed up some sims), but if the use cases are limited, I'd do the hack-the-tcl way for now.

Note, it's actually nice to have two params fed though: a delay before starting waveform dumping and a duration or stop time (duration is usually easier to implement).  Sometimes you want a test to finish for scoreboard checking/fifo draining/etc but you only need the waveforms from a middle part of the sim.

**Comment by tjaychen** [comment]
thanks @sriyerg ! do you want to broadcast this more widely or anything?

**Comment by sriyerg** [comment]
What @cdgori suggested above does not seem too hard to implement (and it was something I already had in mind). Let me see if I can quickly address that.

**Comment by tjaychen** [comment]
> This is slightly tricky, since we are using tcl files to control wave dumping, as opposed to setting up the wave dumping from SV code. If we would have controlled the wave dumping from SV code, then the start time and dump duration could have been easily provided via plusargs. Since we are controlling all that from a tcl file, the only way to pass that control information from outside to the tcl script (which is not invoked standalone, but by the simulator) is via environment variables (AFAIK). dvsim currently does not support setting env vars on a per-test basis. It will be great to hear some thoughts on how to best achieve this.

yeah that was my thinking too. Since it looked like the selection of dump format was also being passed via enviornment variables, i figured start / stop time would be the same. 

Are you suggesting long term we should move the dumping control directly to sv side? I guess in that case we would just need to pass some ifdef's on what simulator we're using and potentially the format right? That sounds good to me, I don't have any objections. However I also don't know if it's something we have to do now, since I'm fine with hacking around it for now :)

**Comment by jdonjdon** [comment]
Alternatively, we can add "-ucli -do {user_dir}/vcs_waves.tcl" after simv. 
As we did -w fsdb.

**Comment by sriyerg** [comment]
Yeah I have been wanting to move the whole wave dump logic to SV since it is a lot easier to control, but I haven't seen clear pros and cons of either approach. 

> Alternatively, we can add "-ucli -do {user_dir}/vcs_waves.tcl" after simv. As we did -w fsdb. For me not just temporal window, but dump hierarchy control also need for chip level sim. Many times, I don't need full depth of tb/*

A slight modification to your suggestion would be to `source` an auxiliary tcl file that would be runtime (i.e. test-specific). It would be sourced after `run 0ns` in sim.tcl so that the user can provide extra dump controls (hierarchies, paths, windows when dumping is enabled etc.) in this file (i.e. this tcl file will only contain dump controls rather than be the full tcl script passed to `-ucli -do switch`).

**Comment by matutem** [comment]
Same test, same seed, with and without dumping:

without dumps:
wall-clock time at start/end: 16:41:26/16:49:23 ==> 7:57

with dumps
wall-clock time at start/end: 16:24:08/16:33:58 ==> 9:50

So dumps takes about 25% more wall clock time. Notice the VCS reported CPU Time is misleading, since it represents the total by all threads, and VCS does all dump IO on a separate thread which runs concurrently with simulation, and may even do other dump event generation outside the main thread.

This is not to dissuade you from doing this, but helps inform you about priorities.

**Comment by matutem** [comment]
This could make much more sense for other purposes, like reducing the generated artifacts that need to be shuffled along, since waves files can get big. It can also make sense for very long running sims that do nothing "interesting" for a long time, but I am sure many of you have had to rerun with more waves when tracing some problem.

**Comment by cdgori** [comment]
That's good data @matutem.

In the past, for the compiled simulators (VCS in particular I know for sure, but NCVerilog/Xcelium I believe as well) if you actually compiled the simulation binary without any waveform dumping linked in at all, you could see a pretty big speedup (versus compiled-in-but-not-dumping where you get the ~20% you saw) because the compilation of the DUT logic could be much more optimized.  The compiler will flatten/not-preserve hierarchy and generally do all the good compiler tricks that it cannot do if hierarchy/internal nodes have to be preserved for waveform dumping.

Unfortunately, I also saw that sometimes tests that failed with waveforms compiled out were hard to reproduce in the waveform enabled sims, probably because of various race conditions in the stimulus gen setups we were using.  This was especially true with full-chip random, which was almost exactly where you would want the speedup, sadly.
